### PR TITLE
#635 StartsRequestTest sometimes fails

### PR DESCRIPTION
**Pull request by maurezen**
Added proxy option to the gpg call. 
Proxy settings are obtained via system http proxy properties.

Note to reviewer:
Somehow the changes from #701 (pull request #713) are shown in this pull request, though #713 is already merged into master.

### CODE REVIEW

**Comment by alex-palevsky** [comment]
Thanks for the pull request, let me find a reviewer..

**Comment by alex-palevsky** [comment]
@yegor256 review this pull request plz

File: src/main/java/com/rultor/agents/req/Decrypt.java [hunk 1]
```diff
@@ -58,11 +59,28 @@
...
+
+    /**
+     * Ctor.
+     * @param prof Profile
+     * @param host Proxy host
+     * @param port Proxy port
+     */
+    Decrypt(final Profile prof, final String host, final int port) {
+        this.profile = prof;
+        this.proxy = this.createProxyString(host, port);
```
Comment (yegor256) [inline src/main/java/com/rultor/agents/req/Decrypt.java]: there should be only one "working" ctor, all others should call it, see http://www.yegor256.com/2014/04/27/typical-mistakes-in-java-code.html#constructors

File: src/main/java/com/rultor/agents/req/Decrypt.java [hunk 2]
```diff
@@ -115,4 +134,18 @@
...
+     * @param host Host.
+     * @param port Port.
+     * @return Command string.
+     */
+    private String createProxyString(final String host, final int port) {
+        return Joiner.on("").join(
+            "http-proxy=",
+            host,
+            ":",
+            port
```
Comment (yegor256) [inline src/main/java/com/rultor/agents/req/Decrypt.java]: can't you put them all in one line? looks like there is enough space

File: src/main/java/com/rultor/agents/req/Decrypt.java [hunk 3]
```diff
@@ -115,4 +134,18 @@
         return commands;
     }
 
+    /**
+     * Proxy usage command creation.
+     * @param host Host.
+     * @param port Port.
+     * @return Command string.
+     */
+    private String createProxyString(final String host, final int port) {
```
Comment (yegor256) [inline src/main/java/com/rultor/agents/req/Decrypt.java]: see http://www.yegor256.com/2014/04/27/typical-mistakes-in-java-code.html#method-names

File: src/main/java/com/rultor/agents/req/StartsRequest.java [hunk 4]
```diff
@@ -173,6 +190,42 @@ public boolean apply(final String input) {
     }
 
     /**
+     * Obtain proxy settings and create a Decrypt instance.
+     * @return Decrypt instance.
+     */
+    private Decrypt createDecrypt() {
```
Comment (yegor256) [inline src/main/java/com/rultor/agents/req/StartsRequest.java]: see http://www.yegor256.com/2014/04/27/typical-mistakes-in-java-code.html#method-names

File: src/main/java/com/rultor/agents/req/StartsRequest.java [hunk 5]
```diff
@@ -173,6 +190,42 @@ public boolean apply(final String input) {
...
+        if (StringUtils.isNotBlank(host)) {
+            final String port = System.getProperty(
+                PORT_KEY,
+                StringUtils.EMPTY
+            );
+            int portValue = DEFAULT_PORT;
+            if (StringUtils.isNotBlank(port)) {
+                try {
+                    portValue = Integer.parseInt(port);
+                } catch (final NumberFormatException exception) {
```
Comment (yegor256) [inline src/main/java/com/rultor/agents/req/StartsRequest.java]: what it's for? why re-throwing? it's an un-checked exception.

**Comment by yegor256** [comment]
@maurezen a few comments above

**Comment by maurezen** [comment]
@yegor256 re-throwing is for message composition, which should imrove log readability. Is entirely optional, though. Do you confirm that this is not needed?

I can't see where Decrypt#createProxyString doesn't conform to the linked method naming guidelines. It doesn't start with 'get' and its name reflects what it does. If it is a violation of the guideline, please do suggest an alternative name or at least specify explicitly what principle is violated. In the meantime, added a javadoc link to Decrypt#proxy.

Extracted obtaining proxy settings from StartsRequest#createDecrypt so it stops violating the naming convention.

Compacted the multiline method call to a single line.

**Comment by yegor256** [comment]
Here is what you do:

```
private Decrypt createDecrypt() {
```

Here is what the article (http://www.yegor256.com/2014/04/27/typical-mistakes-in-java-code.html#method-names) says: _If a method returns something, then its name should explain what it; if it returns void, then its name should explain what it does_

So, your method should be named `decrypt()`. Make sense?

**Comment by maurezen** [comment]
Maybe its just my English, but no, it doesn't. When I first see "decrypt" I think of a verb, not of a noun, hence the disambiguation via *create* prefix. I'll rename it if you insist, but honestly I don't think that is a good idea, since the longer name is not that long.

**Comment by yegor256** [comment]
@maurezen call it "decryptor()" or something like that, but don't start it with a verb if the method returns something. It's conceptually wrong. A method should either update or read. Not both at the same time.

**Comment by maurezen** [comment]
@yegor256 renamed it to #decryptor().

How should one name a factory method in this paradigm?

**Comment by yegor256** [comment]
@maurezen depends on what this factory method returns. If it returns a user, for example, name it `user(int)`

File: src/main/java/com/rultor/agents/req/Decrypt.java [hunk 6]
```diff
@@ -58,11 +67,33 @@
...
+     */
+    private final transient String proxy;
+
+    /**
      * Ctor.
      * @param prof Profile
+     * @param settings Proxy settings container
      */
-    Decrypt(final Profile prof) {
+    public Decrypt(final Profile prof, final Pair<String, Integer> settings) {
```
Comment (yegor256) [inline src/main/java/com/rultor/agents/req/Decrypt.java]: this `public` modifier is redundant, since the class is visible only inside the package

Comment (yegor256) [inline src/main/java/com/rultor/agents/req/Decrypt.java]: why `Pair`? why can't you pass `String` and `int`?

File: src/main/java/com/rultor/agents/req/Decrypt.java [hunk 7]
```diff
@@ -115,4 +147,14 @@
...
     }
 
+    /**
+     * Proxy usage command creation.
+     * @param host Host.
+     * @param port Port.
+     * @return Command string.
+     * @see #proxy
+     */
+    private static String createProxyString(final String host, final int port) {
```
Comment (yegor256) [inline src/main/java/com/rultor/agents/req/Decrypt.java]: same here, don't start with a verb (`create`) if the method is NOT updating the object

**Comment by yegor256** [comment]
@maurezen a few more comments above

**Comment by maurezen** [comment]
@yegor256  I want to get proxy settings in a separate method and a method can only return one value, hence the Pair.

Fixed modifiers.

**Comment by yegor256** [comment]
@maurezen this `Pair` thing looks very wrong. instead of solving your design issues internally, you're putting the problem on the shoulders of the class users. imagine, I will use your class in a few months. and I will see this `Pair` as an argument. I'll be very surprised... at least :) let's refactor this. it's a bad design.

**Comment by maurezen** [comment]
@yegor256 I can see the "least wtf principle" (or YAGNI) as one of the design principles, but not as the sole one, so let's discuss why it is a bad design. 

What I want is to encapsulate the proxy settings in one entity, seeing as proxy host is useless without a proxy port and vice versa, and to decouple data acquisition from Decrypt instance creation. Actually I could write a holder for this pair of values, but 
 - It is introducing a separate entity while we can use an existing one perfectly fine, so 
 - Dependency validator fails as this would 8th data abstraction on which Decrypt is dependant. This could probably be circumvented by uncoupling actual data storage and data access interface, but in the light of the first reason this looks like overengineering squared.

So, is the proxy settings encapsulation really wrong? If yes, why is that?

**Comment by maurezen** [comment]
@yegor256 it's been 36 hours since my commits, and I've seen you answer me at least twice in other threads. Could you please continue the review process?

**Comment by yegor256** [comment]
@maurezen here is how I would do this:

```
Decrypt(Profile prof, String opts) {
  this.profile = prof;
  this.options = opts;
}
Decrypt(Profile prof) {
  this(prof, "");
}
```

Then, in unit tests:

```
new Decrypt(profile, "http-proxy=host:port");
```

In other words, it's not important for the `Decrypt` class what gpg options we're sending. All it needs to know is that we want to add some options to the gpg call. What do you think?

**Comment by maurezen** [comment]
@yegor256 it doesn't look right, either. Suddenly, we have
 - an ability to pass arbitrary option to the call, which is potential point of failure
 - additional pieces of code that are aware of gpg and its option format, which is another point of failure (imagine we want to switch to another tool. We now need to change Decrypt implementation AND all of the constructor calls. Shiny!)

That's why we want to interpret the proxy settings inside Decrypt.

**Comment by yegor256** [comment]
@maurezen fair enough. So, how about this:

```
Decrypt(Profile prof, String host, int port) {
  this.profile = prof;
  if (port == 0) {
    this.options = "";
  } else {
    this.options = String.format("http-proxy=%s:%d", host, port);
  }
}
Decrypt(Profile prof) {
  this(prof, "", 0);
}
```

Then, you get rid of `obtainProxySettings()` and put its code into `decryptor()`.

Besides, try to avoid `StringUtils` usages that often. We don't need their `isBlank` or similar methods because we never work with NULL. So, we don't need to be too protective. Besides, `StringUtils.EMPTY` is also a redundancy. Just `""` is shorter and clear enough.

**Comment by maurezen** [comment]
decryptor() becomes a method doing two distinctly different things at this point, but I could live with that, since we still get the proxy settings in one place.

Do you want StringUtils removed from this pull request, or is that a comment for future consideration?

**Comment by yegor256** [comment]
@maurezen I would suggest to remove it from this PR as much as possible.

**Comment by maurezen** [comment]
@yegor256 done, please take a look.

File: src/main/java/com/rultor/agents/req/Decrypt.java [hunk 8]
```diff
@@ -115,4 +143,14 @@
...
     }
 
+    /**
+     * Proxy usage command creation.
+     * @param host Host.
+     * @param port Port.
+     * @return Command string.
+     * @see #proxy
+     */
+    private static String proxyString(final String host, final int port) {
```
Comment (yegor256) [inline src/main/java/com/rultor/agents/req/Decrypt.java]: why so many lines of code if we can just do `String.format()` right in the constructor? this method looks like over-engineering to me.

File: src/main/java/com/rultor/agents/req/StartsRequest.java [hunk 9]
```diff
@@ -173,6 +187,36 @@ public boolean apply(final String input) {
...
+        Decrypt result;
+        if (host.isEmpty()) {
+            result = new Decrypt(this.profile);
+        } else {
+            final String port = System.getProperty(PORT_KEY, "");
+            int portValue = DEFAULT_PORT;
+            if (!port.isEmpty()) {
+                try {
+                    portValue = Integer.parseInt(port);
+                } catch (final NumberFormatException exception) {
```
Comment (yegor256) [inline src/main/java/com/rultor/agents/req/StartsRequest.java]: what is it for? it's an unchecked exception, why do you catch it here?

File: src/test/java/com/rultor/agents/req/DecryptTest.java [hunk 10]
```diff
@@ -76,6 +77,11 @@
     );
 
     /**
+     * Test profile name.
+     */
+    private static final String TEST_PROFILE_NAME = "test/test";
```
Comment (yegor256) [inline src/test/java/com/rultor/agents/req/DecryptTest.java]: this is a bad idea, see http://www.yegor256.com/2014/04/27/typical-mistakes-in-java-code.html#redundant-constants

File: src/test/java/com/rultor/agents/req/DecryptTest.java [hunk 11]
```diff
@@ -133,4 +134,41 @@ public void decryptsAssets() throws Exception {
...
+    public void testHttpProxyHandling() throws IOException {
+        final Decrypt decrypt = new Decrypt(
+            new Profile.Fixed(
+                this.createTestProfileXML(),
+                TEST_PROFILE_NAME
+            ),
+            "http://someserver.com",
+            8080
+        );
+        final Iterable<String> commands = decrypt.commands();
```
Comment (yegor256) [inline src/test/java/com/rultor/agents/req/DecryptTest.java]: it's a one time variable, see http://www.yegor256.com/2014/04/27/typical-mistakes-in-java-code.html#one-time-variables

**Comment by yegor256** [comment]
@maurezen a few more comments above

**Comment by maurezen** [comment]
@yegor256 
>String-format
Agreed, will be fixed.

>Re-throw
Answered this one already: re-throwing is for message composition, which should imrove log readability. Is entirely optional, though. Do you confirm that this is not needed?

>constant
It is used in more than one place, though. And as soon as someone needs to test something else in Decrypt, it will be used more. Do you confirm that I should inline it in multiple places?

>one-time variable
It reads better than a multiline constructor immediately followed by a method call. Actually, both variables in DecryptTest#testHttpProxyHandling are one-time and could be inlined, but do we really need that? Readability will degrade, while the performance gain is dubious (and while I cannot find anything definite, either compiler or JIT will inline these calls anyway).

**Comment by yegor256** [comment]
I agree about re-throw, let's leave it.

The thing with constant is that it makes an impression to the reader that there is some link between test methods. There is something in common. Why there is nothing. That's why it's better to use unique string literals in every test method.

Yes, one-time variables are bad and should be inlined as much as possible. They are simply adding noise to the code

**Comment by maurezen** [comment]
@yegor256 (not so) fun fact: as soon as I inline something in that method, magic number code style check starts to fail. Will have to introduce a constant.

**Comment by yegor256** [comment]
@maurezen yes, you have to use different string literals in each method

**Comment by maurezen** [comment]
@yegor256 ok, done.

File: src/main/java/com/rultor/agents/req/StartsRequest.java [hunk 12]
```diff
@@ -173,6 +187,36 @@ public boolean apply(final String input) {
...
+                    throw new IllegalStateException(
+                        Joiner.on(' ').join(
+                            "Can't parse proxy port",
+                            port
+                        ),
```
Comment (yegor256) [inline src/main/java/com/rultor/agents/req/StartsRequest.java]: again, why four lines of code when you may do it in one: `String.format("can't parse proxy port %d", port)`. check this http://www.yegor256.com/2014/06/19/avoid-string-concatenation.html

**Comment by yegor256** [comment]
@maurezen to be honest, this decryptor method looks over-complex. I would do it like this:

```
return new Decrypt(
  this.profile, 
  System.getProperty(Decrypt.HOST_KEY, ""), 
  Integer.parseInt(System.getProperty(Decrypt.PORT_KEY, "0"))
);
```

Don't you think?

**Comment by maurezen** [comment]
@yegor256 done.

**Comment by yegor256** [comment]
@rultor try to merge pls

**Comment by maurezen** [comment]
@yegor256 the automatic merge failed. what do we do now?

**Comment by yegor256** [comment]
@rultor try to merge again

**Comment by alex-palevsky** [comment]
@yegor256 I added **46 mins** to your account, many thanks for working with the project!

+46 added to your rating, at the moment it is: [+4937](http://www.netbout.com/b/35008?open=rating)

**Comment by alex-palevsky** [comment]
@rultor deploy now pls

**Pull request merged**